starky/src/prover.rs

```diff
@@ -157,45 +163,57 @@ where
     [(); S::PUBLIC_INPUTS]:,
 {
     let degree = 1 << degree_bits;
-    let points = F::two_adic_subgroup(degree_bits + rate_bits);
 
     // Evaluation of the first Lagrange polynomial on the LDE domain.
     let lagrange_first = {
         let mut evals = PolynomialValues::new(vec![F::ZERO; degree]);
         evals.values[0] = F::ONE;
-        evals.lde(rate_bits)
+        evals.coset_lde(rate_bits)
     };
     // Evaluation of the last Lagrange polynomial on the LDE domain.
     let lagrange_last = {
         let mut evals = PolynomialValues::new(vec![F::ZERO; degree]);
         evals.values[degree - 1] = F::ONE;
-        evals.lde(rate_bits)
+        evals.coset_lde(rate_bits)
     };
 
-    let z_h_on_coset = ZeroPolyOnCoset::new(degree_bits, rate_bits);
+    let z_h_on_coset = ZeroPolyOnCoset::<F>::new(degree_bits, rate_bits);
 
     // Retrieve the LDE values at index `i`.
     let get_at_index = |comm: &PolynomialBatch<F, C, D>, i: usize| -> [F; S::COLUMNS] {
         comm.get_lde_values(i).try_into().unwrap()
     };
+    // Last element of the subgroup.
+    let last = F::primitive_root_of_unity(degree_bits).inverse();
+    let coset = F::cyclic_subgroup_coset_known_order(
+        F::primitive_root_of_unity(degree_bits + rate_bits),
+        F::coset_shift(),
+        degree << rate_bits,
+    );
 
     let quotient_values = (0..degree << rate_bits)
         .into_par_iter()
         .map(|i| {
             // TODO: Set `P` to a genuine `PackedField` here.
             let mut consumer = ConstraintConsumer::<F>::new(
                 alphas.clone(),
+                coset[i] - last,
                 lagrange_first.values[i],
                 lagrange_last.values[i],
             );
             let vars = StarkEvaluationVars::<F, F, { S::COLUMNS }, { S::PUBLIC_INPUTS }> {
                 local_values: &get_at_index(trace_commitment, i),
-                next_values: &get_at_index(trace_commitment, (i + 1) % (degree << rate_bits)),
+                next_values: &get_at_index(
+                    trace_commitment,
+                    (i + (1 << rate_bits)) % (degree << rate_bits),
+                ),
                 public_inputs: &public_inputs,
             };
             stark.eval_packed_base(vars, &mut consumer);
-            // TODO: Fix this once we a genuine `PackedField`.
+            // TODO: Fix this once we use a genuine `PackedField`.
             let mut constraints_evals = consumer.accumulators();
+            // We divide the constraints evaluations by `Z_H(x) / x - last`, i.e., the vanishing
```

Comment:
Hm is this comment from before we settled on adding the `x - last` filter in `ConstraintConsumer`?

Reply:
Oops yes forgot to delete this.